### PR TITLE
(refactor) Generalizing Application Logic - Decoupling from UgandaEMR in Lab Workflows

### PR DESCRIPTION
**Pull request by samuelmale**
## Requirements

- [X] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR primarily removes logic that is specific to UgandaEMR's custom Lab workflows. I'm preparing a follow-up PR to migrate this logic to a dedicated UgandaEMR-specific repository. Additionally, the PR includes bug fixes, general code enhancements. However, there is room for improvement in code quality, and follow-up tasks will be created to address this.

## Screenshots

https://www.loom.com/share/2aa94526398f4114bd96e06d64d56c44
## Related Issue

https://openmrs.atlassian.net/browse/O3-2640

### CODE REVIEW

**Comment by gitcliff** [comment]
LGTM ...

**Comment by gitcliff** [comment]
@pirupius @jabahum @ojwanganto WDYT?

**Review by makombe — COMMENTED** [review]
Thanks @samuelmale  LGTM!

File: src/components/orders-table/orders-data-table.component.tsx [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+                      "No lab requests found. Please check your filters and try again."
+                    )}
+                  </p>
+                </div>
+              </Tile>
+            </div>
+          ) : null}
+          <Pagination
+            forwardText="Next page"
+            backwardText="Previous page"
```
[the remark this answers is not in the document]
Comment (samuelmale) [reply src/components/orders-table/orders-data-table.component.tsx:248]: Yes we do, nice catch!

File: src/components/loader/loader.component.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import React from "react";
+import { useTranslation } from "react-i18next";
+import { InlineLoading } from "@carbon/react";
+import styles from "./loader.scss";
+
+const Loader: React.FC = () => {
```
Comment (ojwanganto) [inline src/components/loader/loader.component.tsx:6]: At some point, I hope we'll have such in a common/core lib where we just reuse.

File: src/lab-tabs/actions/pickup-lab-request-action.component.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    });
+  }, [order]);
+
+  return (
+    <OverflowMenuItem
+      itemText={t("pickupLabRequest", "Pickup Lab Request")}
+      onClick={launchModal}
+      style={{
+        maxWidth: "100vw",
+      }}
```
Comment (pirupius) [inline src/lab-tabs/actions/pickup-lab-request-action.component.tsx:30]: This style seems to be repeated in multiple files. Can we extract this to the styles file so we can have a single point incase of future changes

File: src/lab-tabs/actions/reject-lab-request-action.component.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    });
+  }, [order]);
+
+  return (
+    <OverflowMenuItem
+      itemText={t("rejectLabRequest", "Reject Lab Request")}
+      onClick={launchRejectLabRequestModal}
+      style={{
+        maxWidth: "100vw",
+      }}
```
Comment (pirupius) [inline src/lab-tabs/actions/reject-lab-request-action.component.tsx:29]: The styles will trickle to all action components

File: src/lab-tabs/modals/pickup-lab-request-modal.component.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        mutate();
+        setIsSubmitting(false);
+        closeModal();
+        showSnackbar({
+          isLowContrast: true,
+          title: t("pickedAnOrder", "Picked an order"),
+          kind: "success",
+          subtitle: t(
+            "pickSuccessfully",
+            "You have successfully picked an Order"
```
Comment (pirupius) [inline src/lab-tabs/modals/pickup-lab-request-modal.component.tsx:39]:
```suggestion
            "orderPickedSuccessfully",
            "You have successfully picked an order"
```

File: src/laboratory-resource.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    mutate,
+  };
+}
+
+export function setFulfillerStatus(
+  orderId: string,
+  status: FulfillerStatus,
+  abortController: AbortController
+) {
+  return openmrsFetch(`/ws/rest/v1/order/${orderId}/fulfillerdetails/`, {
```
Comment (pirupius) [inline src/laboratory-resource.ts:41]:
```suggestion
  return openmrsFetch(`{$restBaseUrl}/order/${orderId}/fulfillerdetails/`, {
```

**Review by pirupius — COMMENTED** [review]
Thanks @samuelmale for the amazing work. I've left some comments around some quick fixes we can add to this PR. And i'll start a conversation on the slack channel around the workflows

File: .vscode/settings.json [hunk 7]
```diff
@@ -0,0 +1,2 @@
+{
```
[the remark this answers is not in the document]
Comment (samuelmale) [reply .vscode/settings.json]: This was cleaned up

File: src/config-schema.ts [hunk 8]
```diff
@@ -1,53 +1,20 @@
...
-    _default: "b1f8b6c8-c255-4518-89f5-4236ab76025b",
-    _description: "Concept uuid for laboratory referals destinations",
-  },
-
-  enableSendingLabTestsByEmail: {
-    _type: Type.Boolean,
-    _default: false,
-    _description: "This enables sending results to patient via email",
-  },
-  enableSpecimenIdAutoGeneration: {
```
[the remark this answers is not in the document]
Comment (samuelmale) [reply src/config-schema.ts:40]: @pirupius Do we have a clear ID generation strategy? I checked the [KenyaEMR instance](https://dev.kenyahmis.org/openmrs/spa/home/billing) and this feature doesn't exist and I believed it's one of the reasons they forked

File: src/lab-tabs/actions/pickup-lab-request-action.component.tsx (hunk 3 above)
Comment (samuelmale) [reply src/lab-tabs/actions/pickup-lab-request-action.component.tsx:30]: Well like I said, there lots of residue snippets that don't align with our standard practices, this is one of them. More work is required to harmonise the styling in the entire repo. This was just to get us started

**Comment by samuelmale** [comment]
The "Tests ordered" tab is like a bucket of all tests ordered with different statuses. It's equipped with a filter that lets you filter the list by status. However there was no way for say "view new orders; orders without a fulfiller status and with a new action". To solve this problem, I've introduced a filter option called "NEW". This is important for UgandEMR's workflow.

**Comment by ojwanganto** [comment]
@samuelmale I don't understand the rationale for the new filter especially for statuses already covered in other tabs. We need to discuss the transition process and the corresponding tabs that various fulfiller actions should map to. Am certain lab processes are common across different countries/implementations

**Comment by samuelmale** [comment]
@ojwanganto The "Tests ordered" is like a bucket of orders with different fulfiller statuses. However this varies across implementations, eg. For UgandaEMR's use case this tab list only orders freshly ordered without a fulfiller status. This ensures that the lab attendants focus on the incoming fresh orders to take action. 
> I don't understand the rationale for the new filter especially for statuses already covered in other tabs.

Imagine having a very long mixed (orders of disparate statuses) list of orders but without a way of applying a filter that will let you focus on the recently published orders? The main rationale for this filter option (a derived status) is to bring in the ability of viewing those new orders. However, if this seems unnecessary I can revert; just let me know

**Comment by pirupius** [comment]
@samuelmale @ojwanganto this takes us back to the [conversation i started on slack](https://openmrs.slack.com/archives/C061LK7AS8H/p1712057139898299) about having that bucket of statuses (aka NEW) being the only ones shown in the first tab. It's the landing point and the primary users for this are the lab attendants who want to know what lab orders need to be worked on.

**Comment by ojwanganto** [comment]
Thanks @samuelmale, @pirupius - my guess is that the implementation of the first tab should be common across :). It is basically what is fresh from a clinician to the lab.

**Comment by samuelmale** [comment]
@pirupius @ojwanganto I'm on board with dropping the filter in the "Tests Ordered" table and default to loading "NEW" orders. However this idea of having the filter was something the DIGI team was interested in and I believed we had arrived at a consensus of us having it that way?  Should we drop the filter? cc: @gitcliff @caseyi @slubwama @jabahum

**Comment by jabahum** [comment]
LGTM unless there's any concerns from others

**Review by jabahum — COMMENTED** [review]
LGTM

**Comment by pirupius** [comment]
Merging this in to unblock other tasks. Any other enhancements will come in follow up PR's

**Pull request merged**